### PR TITLE
8307466: java.time.Instant calculation bug in until and between methods

### PR DESCRIPTION
**Pull request by RogerRiggs**
The implementation of java.time.Instant.until(I2, ChronoUnit) in some cases did not correctly borrow or carry from the nanos to the seconds when computing using ChronoUnit.MILLIS or ChronoUnit.MICROS.
The errant computation was introduced by [JDK-8273369](https://bugs.openjdk.org/browse/JDK-8273369).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8307578](https://bugs.openjdk.org/browse/JDK-8307578) to be approved

### Issues
 * ⚠️ Temporary failure when trying to retrieve information on issue `8307466`.
 * [JDK-8307578](https://bugs.openjdk.org/browse/JDK-8307578): Error computing the amount of milli and micro seconds between java.time.Instants (**CSR**)

### Reviewers
 * [Stephen Colebourne](https://openjdk.org/census#scolebourne) (@jodastephen - Author) ⚠️ Review applies to [8deb8df7](https://git.openjdk.org/jdk/pull/13846/files/8deb8df737950d1ddf76c785ca94e2b0b0d54520)
 * [Naoto Sato](https://openjdk.org/census#naoto) (@naotoj - **Reviewer**)

### Contributors
 * Raffaello Giulietti `<rgiulietti@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13846/head:pull/13846` \
`$ git checkout pull/13846`

Update a local copy of the PR: \
`$ git checkout pull/13846` \
`$ git pull https://git.openjdk.org/jdk.git pull/13846/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13846`

View PR using the GUI difftool: \
`$ git pr show -t 13846`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13846.diff">https://git.openjdk.org/jdk/pull/13846.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13846#issuecomment-1536962213)

### CODE REVIEW

**Comment by RogerRiggs** [comment]
/contributor rgiulietti  Contributed the corrected implementation.

**Review by jodastephen — APPROVED** [review]

**Comment by RogerRiggs** [comment]
/csr needed
The change in behavior should be documented and release noted

File: src/java.base/share/classes/java/time/Instant.java [hunk 1]
```diff
@@ -1169,15 +1169,25 @@ private long nanosUntil(Instant end) {
     }
 
     private long microsUntil(Instant end) {
-        long secsDiff = Math.subtractExact(end.seconds, seconds);
-        long totalMicros = Math.multiplyExact(secsDiff, MICROS_PER_SECOND);
-        return Math.addExact(totalMicros, (end.nanos - nanos) / 1000);
+        long microsDiff = Math.multiplyExact(end.seconds - seconds, MICROS_PER_SECOND);
+        long nanosDiff = end.nanos - nanos;
```
Comment (rgiulietti) [inline src/java.base/share/classes/java/time/Instant.java:1173]: FWIW, `nanosDiff` could be declared `int` to slightly speed up the division later in the code, while still not risking overflows.

Similarly in `millisUntil()`.

**Comment by RogerRiggs** [comment]
Please review the CSR. [8307578](https://bugs.openjdk.org/browse/JDK-8307578)
Suggestions on a clear description are appreciated.

**Comment by rgiulietti** [comment]
I think that the explanation of the old code failing to properly consider carries and borrows, in the "Problem" section, is quite clear.

For the record, I added a more math-oriented comment.

**Comment by RogerRiggs** [comment]
@rgiulietti Thank you for the detailed math-oriented explanation.

**Pull request closed without merging**